src/super_gradients/training/datasets/balancing_classes_utils.py

```diff
@@ -0,0 +1,88 @@
+import math
+from typing import Callable, List, Optional
+
+import numpy as np
+from torch.utils.data import Dataset
+
+
+def _default_oversample_heuristic(cat_id, cat_freq, oversample_threshold):
+    """
+    How to oversample a category. This heuristic is based on https://arxiv.org/pdf/1908.03195.pdf.
+    A value of 1.0 corresponds to no oversampling.
+    """
+    return max(1.0, math.sqrt(oversample_threshold / cat_freq))
+
+
+def get_repeat_factors(
+    *,
+    index_to_classes: Callable[[int], List[int]],
+    num_classes: int,
+    dataset_length: int,
+    ignore_empty_annotations: bool = False,
+    oversample_threshold: Optional[float] = None,
+    oversample_heuristic: Optional[Callable[[int, int, float], float]] = _default_oversample_heuristic,
+) -> List[float]:
+    """
+    Oversampling scarce classes from detection dataset, following sampling strategy described in https://arxiv.org/pdf/1908.03195.pdf.
+    Implementation based on https://github.com/open-mmlab/mmdetection (Apache 2.0 license).
+
+    :param index_to_classes: A function that maps an index to a list of classes (labels).
+    :param num_classes: Number of classes.
+    :param dataset_length: Length of the dataset.
+    :param ignore_empty_annotations: Whether to ignore empty annotations.
+    :param oversample_threshold: A frequency threshold (fraction, 0-1). Classes that are *less frequent* than this threshold will be oversampled.
+                                    Default: None. If None, the median of the class frequencies will be used.
+                                    See step (1) below to understand how frequencies are computed.
+
+
+    The repeat factor is computed as followed.
+    1. For each category c, compute the fraction # of images that contain it (its frequency): :math:`f(c)`
+    2. For each category c, compute the category-level repeat factor: :math:`r(c) = max(1, sqrt(t/f(c)))`
+    3. For each image I, compute the image-level repeat factor: :math:`r(I) = max_{c in I} r(c)`
+
+    Returns a list of repeat factors (length = dataset_length). How to read: result[i] is a float, indicates the repeat factor of image i.
```

Comment:
Out of curiosity - I guess  this would mean that using Mosaic / mixup would mess things up here?
We use them in most of our recipes, so maybe there's something to tak in account here.

Reply:
So, you are half correct :) I believe things should work ok because the balancing is done (currently) in the sampler-level, while mixup/mosaic are dataset-level. On the other hand, the sampler will "balance" and ask for more indices with scarce classes (this is good), but the dataset does not know that it should sample non uniformly, thus 3/4 images will be taken at random, regardless of class scarcity. Hope it makes sense...